WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainState.kt

```diff
@@ -0,0 +1,274 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.android.wearable.speaker
+
+import android.Manifest
+import android.app.Activity
+import android.content.pm.PackageManager
+import android.media.AudioDeviceInfo
+import android.media.AudioManager
+import android.media.MediaPlayer
+import androidx.annotation.RequiresPermission
+import androidx.compose.foundation.MutatorMutex
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.core.content.ContextCompat
+import androidx.core.content.getSystemService
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import java.time.Duration
+import kotlin.coroutines.resume
+
+/**
+ * A state holder driving the logic of the app.
+ *
+ * This state holder is scoped to the composition (and therefore can be provided an [Activity]) to handle
+ * permission-related logic.
+ */
+class MainState(
+    private val activity: Activity,
+    private val requestPermission: () -> Unit,
+    private val showPermissionRationale: () -> Unit,
+    private val showSpeakerNotSupported: () -> Unit
+) {
+    /**
+     * The [MutatorMutex] that guards the state of the app.
+     *
+     * Due to being a [MutatorMutex], this automatically handles cleanup of any ongoing asynchronous work,
+     * like playing music or recording, ensuring that only one operation is occurring at a time.
+     */
+    private val appStateMutatorMutex = MutatorMutex()
+
+    /**
+     * The primary playback state.
+     */
+    var playbackState by mutableStateOf<PlaybackState>(PlaybackState.Ready)
+        private set
+
+    /**
+     * The progress of an ongoing recording.
+     *
+     * Note that this value can be read even when recording is not occuring, in which case it corresponds to the last
+     * known value of recording progress (or 0), where that value is useful for animations.
+     */
+    var recordingProgress by mutableStateOf(0f)
+        private set
+
+    /**
+     * `true` if we know the user has denied the record audio permission.
+     */
+    var isPermissionDenied by mutableStateOf(false)
+        private set
+
+    /**
+     * The [SoundRecorder] for recording and playing audio captured on-device.
+     */
+    private val soundRecorder = SoundRecorder(activity, "audiorecord.pcm")
+
+    suspend fun onStopped() {
+        appStateMutatorMutex.mutate {
+            playbackState = PlaybackState.Ready
+        }
+    }
+
+    suspend fun onMicClicked() {
+        appStateMutatorMutex.mutate {
```

Comment:
Holding the mutex at this level seems unsafe. Given that in this block you call suspend functions (e.g. `record` in the first block), it might be the case that you hold the mutex for too long. Is that the intention? Is the intention that while `onMicClicked` is being executed, other functions cannot update the state?

Maybe the functions need to be called differently? `onMicClicked` is not self-explanatory, it just tell us that it's handling a user event. Then I have to go through the body to actually see what's going on.

Reply:
On the other hand, it makes sense that the functions are named based on the input they're processing as this is a state holder that's tightly coupled to the UI. Up to you!